ratelimit/src/lib.rs

```diff
@@ -1,203 +1,293 @@
+        }
+
+        // figure out how many tokens we might add
+        let amount = intervals * refill_amount;
+
+        let available = self.available.load(Ordering::Acquire);
+        let capacity = self.capacity.load(Ordering::Relaxed);
+
+        if available + amount >= capacity {
+            self.available
+                .fetch_add(capacity - available, Ordering::Release);
+        } else {
+            self.available.fetch_add(amount, Ordering::Release);
+        }
+
+        Ok(())
     }
 
-    // internal function to move the time forward
-    fn tick(&self) {
-        let now = Instant::<Nanoseconds<u64>>::now();
-        let next = self.next.load(Ordering::Relaxed);
-        if now >= next {
-            let strategy = Refill::try_from(self.strategy.load(Ordering::Relaxed));
-            let tick = match strategy {
-                Ok(Refill::Smooth) => self.tick.load(Ordering::Relaxed).as_nanos(),
-                Ok(Refill::Uniform) => self.uniform.sample(&mut rand::thread_rng()) as u64,
-                Ok(Refill::Normal) => self.normal.sample(&mut rand::thread_rng()) as u64,
-                Err(_) => self.tick.load(Ordering::Relaxed).as_nanos(),
-            };
+    /// Non-blocking function to "wait" for a single token. On success, a single
+    /// token has been acquired. On failure, a `Duration` hinting at when the
+    /// next refill would occur is returned.
+    pub fn try_wait(&self) -> Result<(), core::time::Duration> {
+        let refill_result = self.refill(Instant::now());
+
+        loop {
+            let available = self.available.load(Ordering::Acquire);
+            if available == 0 {
+                refill_result?
+            }
+
+            let new = available - 1;
+
             if self
-                .next
-                .compare_exchange(
-                    next,
-                    next + Duration::<Nanoseconds<u64>>::from_nanos(tick),
-                    Ordering::SeqCst,
-                    Ordering::SeqCst,
-                )
+                .available
+                .compare_exchange(available, new, Ordering::AcqRel, Ordering::Acquire)
                 .is_ok()
             {
-                let quantum = self.quantum.load(Ordering::Relaxed);
-                let capacity = self.capacity.load(Ordering::Relaxed);
-                let available = self.available.load(Ordering::Relaxed);
-                if available + quantum >= capacity {
-                    let quantum = capacity - available;
-                    self.available.fetch_add(quantum, Ordering::Relaxed);
-                } else {
-                    self.available.fetch_add(quantum, Ordering::Relaxed);
-                }
+                return Ok(());
             }
         }
     }
+}
 
-    /// Non-blocking wait for one token, returns an `Ok` if a token was
-    /// successfully acquired, returns an `Err` if it would block.
-    ///
-    /// # Examples
+pub struct Builder {
+    available: u64,
```

Comment:
I think `initial_capacity` is a better name, and save `available` for internal use only in `refill`/`try_wait` implementations.

Reply:
See latest commit and see if the names look better now.